plugin/rollback.go

```diff
@@ -0,0 +1,75 @@
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/mitchellh/mapstructure"
+)
+
+const (
+	rotateCredentialWAL = "rotateCredentialWAL"
+)
+
+// rotateCredentialWAL is used to store information in a WAL that can retry a
+// credential rotation in the event of partial failure.
+type rotateCredentialEntry struct {
+	LastVaultRotation  time.Time `json:"last_vault_rotation"`
+	LastPassword       string    `json:"last_password"`
+	CurrentPassword    string    `json:"current_password"`
+	RoleName           string    `json:"name"`
+	ServiceAccountName string    `json:"service_account_name"`
+	TTL                int       `json:"ttl"`
+	walID              string
+}
+
+func (b *backend) walRollback(ctx context.Context, req *logical.Request, kind string, data interface{}) error {
+	switch kind {
+	case rotateCredentialWAL:
+		return b.handleRotateCredentialRollback(ctx, req.Storage, data)
+	default:
+		return fmt.Errorf("unknown WAL entry kind %q", kind)
+	}
+}
+
+func (b *backend) handleRotateCredentialRollback(ctx context.Context, storage logical.Storage, data interface{}) error {
+	var wal rotateCredentialEntry
+	if err := mapstructure.WeakDecode(data, &wal); err != nil {
+		return err
+	}
```

Comment:
Hmm, I'm wondering if we can always assume a rollback is needed when we get a WAL entry here. I think there are scenarios where we will want to just discard it. For example, if the entire operation in `generateAndReturnCreds()` is successful besides the final deletion of the WAL, the user will think the operation was successful. It may surprise them to see that the creds were eventually rolled back.

You may be able to read the current state of creds from storage + the WAL to make the proper decision.

Reply:
I added some logic for this.